rad/prelude/patterns.rad

```diff
@@ -0,0 +1,192 @@
+  (fn [m n]
+    (def excl-merge (exclusive-dict-merge m n))
+    (def cs (lookup :conflicts excl-merge))
+    (def merge (lookup :merge excl-merge))
+    (if (eq? cs {})
+      [:Just merge]
+      (maybe-foldlM (fn [acc kv]
+                      (def v (nth 1 kv))
+                      (def v0 (nth 0 v))
+                      (if (eq? v0 (nth 1 v))
+                        [:Just (insert (nth 0 kv) v0 acc)]
+                        :Nothing))
+                    merge
+                    (seq cs)))))
+(def _
+  "The wildcard pattern."
+  (fn [v] [:Just {}]))
+
+(def-rec match-pat
+  "The pattern matching dispatch function. This function defines how patterns
+  are treated in `match` expressions. Atoms are treated as bindings. Numbers,
+  keywords and strings are constant patterns. Dicts of patterns match dicts
+  whose values at those keys match those patterns. Vectors of patterns match
+  vectors of the same length, pairing the patterns and elements by index."
+  (fn [pat v]
+    (def /vec
+      (fn [pats]
+        (fn [v]
+          (if (eq? (length v) (length pat))
+            (maybe-foldlM
+             (fn [acc pair]
+               (maybe->>= (match-pat (nth 0 pair) (nth 1 pair))
+                          (fn [bs] (non-linear-merge acc bs))))
+             {}
+             (zip pats v))
+            :Nothing))))
+    (def /dict
+      (fn [keys-pat]
+        (fn [d]
+          (maybe-foldlM
+           (fn [acc kv]
+             (def k (nth 0 kv))
+             (def pat (nth 1 kv))
+             (if (member? k d)
+               (maybe->>= (match-pat pat (lookup k d))
+                          (fn [bs] (non-linear-merge acc bs)))
+               :Nothing))
+           {}
+           (seq keys-pat)))))
+    (cond
+      (atom? pat)
+      [:Just {pat v}]
+      (some [(number? pat) (keyword? pat) (string? pat)])
+      (if (eq? pat v) [:Just {}] :Nothing)
+      (dict? pat)
+      ((/dict pat) v)
+      (vector? pat)
+      ((/vec pat) v)
+      :else
+      (pat v))))
+
+(:test "basic pattern matching"
+       [ (match 42 12 :not-42 'x x) ==> 42 ]
+       [ (match 42 12 :not-42 42 :yes) ==> :yes ]
+       [ (match :fourty-two :not-42 :no :fourty-two :yes) ==> :yes ]
+       [ (match "42" "?" :no "42" :yes) ==> :yes ]
+       [ (match {:a :a :b :b} {:a 'a :b 'b} [a b]) ==> [:a :b] ]
+       [ (match [1 2 3] ['x _ 'y] [x y]) ==> [1 3] ]
+       [ (match {:a 1 :b [1 2 3]}
+                42 :no
+                ({:a _ :b [_ 'n _]} | (eq? n 2)) :yes
+                _ :no) ==> :yes ]
+       [ (match [1 2]
+                ['x 'x] :no
+                ['x 'y] :yes) ==> :yes ]
+       [ (match [1 1]
+                ['x 'x] :yes
+                ['x 'y] :no) ==> :yes])
+
+(def /Just
+  "Pattern which matches `[:Just x]`."
+  (fn [pat]
+    (fn [v]
+      (maybe->>= v
+                 (fn [x] (match-pat pat x))))))
+
+(:test "Just patterns"
+       [ (match [:Just 42] (/Just 'x) x _ :no) ==> 42 ]
```

Comment:
I'm still not clear on why we've settled on `:Just`/`:Nothing` over `:just`/`:nothing`, but since we have already, I guess we can talk about that in some other occasion.

Reply:
Actually, a related question is the syntax for pattern functions. I have chosen a prefix `\` here, but I also considered using capital letters: `Just`, `Cons`, etc. This would make it consistent with the use of capital letters when mimicking "constructors" in Haskell.

Reply:
I slightly prefer titlecase, but up to you.